q2_annotate/kraken2/merge.py

```diff
@@ -0,0 +1,420 @@
+                    'reports that contain minimizer information.'
+                )
+                raise ValueError(msg)
+            else:
+                mapping[sample_id].append(format)
+
+    report_mapping = {}
+    for sample_id, value in chained_reports:
+        _update_mapping(
+            sample_id, value, report_mapping, Kraken2ReportFormat
+        )
+
+    output_mapping = {}
+    for sample_id, value in chained_outputs:
+        _update_mapping(
+            sample_id, value, output_mapping, Kraken2OutputFormat
+        )
+
+    return report_mapping, output_mapping
+
+
+def _check_for_minimizers(reports: Kraken2ReportDirectoryFormat) -> bool:
+    '''
+    Checks whether the reports being processed include the optional minimizer
+    columns. This determines whether reports with a shared sample ID can be
+    merged.
+
+    Parameters
+    ----------
+    reports : Kraken2ReportDirectoryFormat
+        The kraken2 reports to examine for minimzer columns.
+
+    Returns
+    -------
+    bool
+        Wether minimizer columns are present in the reports.
+    '''
+    _, report = list(reports.reports.iter_views(Kraken2ReportFormat))[0]
+
+    return 'n_read_minimizers' in report.view(pd.DataFrame)
+
+
+def _merge_reports(
+    reports: list[Kraken2ReportFormat]
+) -> Kraken2ReportFormat:
+    '''
+    Merges two or more kraken2 reports into a single report. See
+    `_merge_trees` for the actual merging algorithm.
+
+    Parameters
+    ----------
+    reports : list[Kraken2ReportFormat]
+        Kraken2 reports belonging to the same sample ID.
+
+    Returns
+    -------
+    Kraken2ReportFormat
+        The merged kraken2 report format.
+    '''
+    if len(reports) == 1:
+        return reports[0]
+
+    merged: tuple[TreeNode | None, TreeNode | None] | None = None
+    while reports:
+        first = _report_df_to_tree(reports.pop().view(pd.DataFrame))
+        if merged is None:
+            second = _report_df_to_tree(reports.pop().view(pd.DataFrame))
+        else:
+            second = merged
+
+        merged = _merge_trees(first, second)
+
+    merged_report_df = _dump_tree_to_report(*merged)
+    merged_report = Kraken2ReportFormat()
+    merged_report_df.to_csv(
+        str(merged_report), sep='\t', header=False, index=False
+    )
+
+    return merged_report
+
+
+def _merge_outputs(
+    outputs: list[Kraken2OutputFormat]
+) -> Kraken2OutputFormat:
+    '''
+    Merges two or more kraken2 outputs into a single output. Output files
+    are merged by concatenation.
+
+    Parameters
+    ----------
+    outputs : list[Kraken2OutputFormat]
+        Kraken2 outputs belonging to the same sample ID.
+
+    Returns
+    -------
+        The merged kraken2 output format.
+    '''
+    if len(outputs) == 1:
+        return outputs[0]
+
+    merged_output = Kraken2OutputFormat()
+    while outputs:
+        with (
+            open(str(merged_output), 'a') as merged_fh,
```

Comment:
I would open this once, outside of the while, then you don't need the somewhat unusual mode `'a'` and we can keep the same file descriptor around.